webapp/views/launch/script.js

```diff
@@ -528,20 +528,34 @@ function buttonElement(action) {
  * @param {string} className className of the table
  */
 function tableElement(className, data) {
-  let html = `<table class=${className}><tr>`
-  Object.keys(data[0]).forEach(field => {
-    html += `<th>${field}</th>`
-  })
-  html += '</tr>'
-  data.forEach(entry => {
-    html += '<tr>'
-    Object.keys(entry).forEach(field => {
-      html += `<td>${entry[field]}</td>`
+  const table = $(`<table class=${className}><tr>`)
+  const headRow = $(`<tr></tr>`)
+
+  table.append(headRow)
+
+  if (Array.isArray(data)) {
```

Comment:
That's why I love TypeScript. If the function signature were `tableElement(className: string, data: any[])`, this would not work!
Also there is too much repetition here. Lines 537-539 are repeated in lines 549ff. and lines 541-546 are repeated in lines 552ff. Try to refactor so this gets eliminated.

Reply:
Could you be more explicit on this one?

Reply:
Yes, exactly. The problem here is that the `data` argument is ambiguous: you allow it to be a simple object or an array of objects. But actually what you need for the table is a header row and multiple content rows. So why not write a function that takes `headFields` (array of strings) and `rows` (array of objects containing values) as arguments? Then the body of the function will look like:
```
headFields.forEach(field => {
  headRow.append($(`<th>${field}</th>`))
})
rows.forEach(entry => {
  const contentRow = $(`<tr></tr>`)
  table.append(contentRow)
  Object.values(entry).forEach(value => {
    contentRow.append($(`<td>${value}</td>`))
  })
})
```
Much cleaner!

Reply:
Agree! Thanks a lot ;)